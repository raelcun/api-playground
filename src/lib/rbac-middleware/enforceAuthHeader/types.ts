export type Config = { jwtSecret: string }
export type ConfigProvider = () => Config
