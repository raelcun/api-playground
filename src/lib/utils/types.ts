import * as t from 'io-ts'

interface NonEmptyStringBrand {
  readonly NonEmptyString: unique symbol
}

export const NonEmptyString = t.brand(
  t.string,
  (s): s is t.Branded<string, NonEmptyStringBrand> => s !== undefined && s.length > 0,
  'NonEmptyString',
)

export type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>
}

export type ChangeTypeOfKeys<T, Keys extends keyof T, NewType> = {
  [key in keyof T]: key extends Keys ? NewType : T[key]
}

export type ArgumentTypes<T> = T extends (...args: infer U) => infer R ? U : never

export type ReplaceReturnType<T, TNewReturn> = (...a: ArgumentTypes<T>) => TNewReturn
