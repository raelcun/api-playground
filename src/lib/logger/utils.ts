import { either as E, taskEither as TE } from 'fp-ts'

import { Err } from '@lib/error'

import { Logger, LogMethods } from './types'

export const logErrorsE = <L extends Err<string>>(logger: Logger, method: LogMethods = 'trace') =>
  E.mapLeft<L, L>(e => {
    logger.log({ level: method, payload: e })
    return e
  })

export const logErrorsTE = <L extends Err<string>>(logger: Logger, method: LogMethods = 'trace') =>
  TE.mapLeft<L, L>(e => {
    logger.log({ level: method, payload: e })
    return e
  })
