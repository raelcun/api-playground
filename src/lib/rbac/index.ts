export * from './types'
export { getEnforcer, enforceRole } from './enforcer'
