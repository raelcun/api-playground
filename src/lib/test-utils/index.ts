import * as HttpStatus from 'http-status-codes'
import { Context } from 'koa'

import { Logger } from '@lib/logger'
import { install, InstalledClock } from '@sinonjs/fake-timers'

export const createMockLogger = (): Logger => ({
  log: jest.fn(),
  transform: jest.fn(),
  defaultValues: jest.fn(),
  pinValues: jest.fn(),
  child: jest.fn(),
})

/* istanbul ignore next */
export const installStaticClock = () => {
  let clock: InstalledClock

  beforeAll(() => {
    clock = install({ now: 946684800000 })
  })

  afterAll(() => {
    clock.uninstall()
  })

  return () => clock
}
