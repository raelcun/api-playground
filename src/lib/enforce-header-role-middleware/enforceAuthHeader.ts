import { either as E, taskEither as TE } from 'fp-ts'
import { flow } from 'fp-ts/lib/function'
import { pipe } from 'fp-ts/lib/pipeable'
import HttpStatus from 'http-status-codes'
import * as t from 'io-ts'
import { Middleware } from 'koa'

import { Err } from '@lib/error'
import { logErrorsTE, LoggerProvider } from '@lib/logger'
import { Actions, EnforceProvider, enforceRole, Roles } from '@lib/rbac'
import { createMiddlewareTE } from '@lib/utils'

import { resolveAuthHeader, verifyAndParseToken } from './parseAuthHeader'
import { ConfigProvider } from './types'

const enforceWithAuthHeaderInternal = <T>(
  enforceProvider: EnforceProvider,
  configProvider: ConfigProvider,
  tokenV: t.Type<T, unknown>,
  getRole: (token: T) => Roles,
) => <T extends keyof Actions, U extends Actions[T]>(resource: T, actions: U[]) => (
  headers: unknown,
): TE.TaskEither<Err, void> =>
  pipe(
    resolveAuthHeader(headers),
    flow(E.chain(verifyAndParseToken(configProvider, tokenV)), TE.fromEither),
    TE.chain(token => enforceRole(enforceProvider)(resource)(actions)(getRole(token))),
  )

export const createEnforceWithAuthHeaderMiddleware = <T>(
  enforceProvider: EnforceProvider,
  configProvider: ConfigProvider,
  getLogger: LoggerProvider,
  tokenV: t.Type<T, unknown>,
  getRole: (token: T) => Roles,
) => <T extends keyof Actions, U extends Actions[T]>(resource: T, actions: U[]): Middleware =>
  createMiddlewareTE(ctx =>
    pipe(
      enforceWithAuthHeaderInternal(
        enforceProvider,
        configProvider,
        tokenV,
        getRole,
      )(
        resource,
        actions,
      )(ctx.headers),
      logErrorsTE(getLogger(), 'error'),
      TE.mapLeft(() => {
        ctx.status = HttpStatus.UNAUTHORIZED
      }),
    ),
  )
