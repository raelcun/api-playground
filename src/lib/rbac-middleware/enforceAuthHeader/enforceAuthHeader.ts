import { either as E, taskEither as TE } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'
import HttpStatus from 'http-status-codes'
import { Middleware } from 'koa'

import { Err } from '@lib/error'
import { Actions, EnforceProvider, enforceRole } from '@lib/rbac'
import { createMiddlewareTE } from '@lib/utils'

import { resolveAuthHeader, verifyAndParseToken } from './parseAuthHeader'
import { ConfigProvider } from './types'

const enforceWithAuthHeaderInternal = (enforceProvider: EnforceProvider) => (
  configProvider: ConfigProvider,
) => <T extends keyof Actions, U extends Actions[T]>(resource: T, actions: U[]) => (
  headers: unknown,
): TE.TaskEither<Err, void> =>
  pipe(
    resolveAuthHeader(headers),
    E.chain(verifyAndParseToken(configProvider)),
    TE.fromEither,
    TE.chain(({ role }) => enforceRole(enforceProvider)(resource)(actions)(role)),
  )

export const createEnforceWithAuthHeaderMiddleware = (enforceProvider: EnforceProvider) => (
  configProvider: ConfigProvider,
) => <T extends keyof Actions, U extends Actions[T]>(resource: T, actions: U[]): Middleware =>
  createMiddlewareTE(ctx =>
    pipe(
      enforceWithAuthHeaderInternal(enforceProvider)(configProvider)(resource, actions)(
        ctx.headers,
      ),
      TE.mapLeft(() => {
        ctx.status = HttpStatus.UNAUTHORIZED
      }),
    ),
  )
