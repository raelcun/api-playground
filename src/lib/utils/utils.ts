import { array as A, either as E, io as IO, taskEither as TE } from 'fp-ts'
import { pipe } from 'fp-ts/lib/pipeable'
import * as t from 'io-ts'
import { reporter } from 'io-ts-reporters'
import { Context, Middleware } from 'koa'

import { Err } from '@lib/error'

export const tryCatchTE: typeof TE.tryCatch = (f, onRejected) =>
  TE.tryCatch(() => {
    try {
      return f()
    } catch (err) {
      return Promise.reject(err)
    }
  }, onRejected)
tryCatchTE.call
export const createVoidTE = <L>() => TE.rightTask<L, void>(async () => {})

export const filterObjectKeys = <T extends Record<string, unknown>, K extends keyof T>(
  o: T,
  keys: K[],
): Pick<T, K> => {
  const res: any = {}

  keys.forEach(key => {
    res[key] = o[key]
  })

  return res
}

export const mapErrorCode = <T extends string>(newErrorCode: T) =>
  E.mapLeft((originalError: Err<string>) => ({
    ...originalError,
    code: newErrorCode,
    subcode: originalError.code,
  }))

export const decode = <T>(
  t: t.Type<T, unknown>,
  v: unknown,
): E.Either<Err<'VALIDATION_ERROR'>, T> =>
  pipe(
    t.decode(v),
    E.mapLeft(errors => ({
      code: 'VALIDATION_ERROR',
      message: reporter(E.left(errors)).join('. '),
    })),
  )

export const ioSequence = A.array.sequence(IO.io)

export const DateFromString = new t.Type<Date, string, unknown>(
  'DateFromString',
  (u): u is Date => u instanceof Date,
  (u, c) =>
    pipe(
      t.union([t.string, t.number]).validate(u, c),
      E.chain(s => {
        const d = new Date(s)
        return isNaN(d.getTime()) ? t.failure(u, c) : t.success(d)
      }),
    ),
  a => a.toISOString(),
)

export const createMiddlewareTE = <T>(
  f: (ctx: Context) => TE.TaskEither<unknown, unknown>,
): Middleware<T> => async (ctx, next) => {
  await pipe(
    f(ctx),
    TE.chain(() => TE.rightTask(next)),
  )()
}

export const createMiddlewareE = <T>(
  f: (ctx: Context) => E.Either<unknown, unknown>,
): Middleware<T> => createMiddlewareTE(ctx => TE.fromEither(f(ctx)))
