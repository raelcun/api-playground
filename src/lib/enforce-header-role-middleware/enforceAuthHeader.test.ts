import { taskEither as TE } from 'fp-ts'
import HttpStatus from 'http-status-codes'
import * as t from 'io-ts'
import { sign } from 'jsonwebtoken'
import { Context } from 'koa'

import { EnforceProvider, Roles, RolesV } from '@lib/rbac'
import { createMockLogger } from '@lib/test-utils'
import { createKoaContext, createMockNext } from '@modules/utils'

import { createEnforceWithAuthHeaderMiddleware } from './enforceAuthHeader'

const getConfig = () => ({ jwtSecret: 'foobar' })

const createMockContextAndNext = (authHeader: string) => ({
  mockContext: createKoaContext({
    headers: { authorization: authHeader },
  }),
  mockNext: createMockNext(),
})

const mockTokenV = t.type({
  userId: t.string,
  role: RolesV,
})
type MockToken = t.TypeOf<typeof mockTokenV>
const getRoleFromToken = (token: MockToken) => token.role

const createExpiredTokenHeader = () =>
  `Bearer ${sign(<MockToken>{ userId: 'foo', role: Roles.User }, getConfig().jwtSecret, {
    expiresIn: '-10s',
  })}`

const createInvalidTokenSchemaHeader = () =>
  `Bearer ${sign(<MockToken>{ userId: 'foo' }, getConfig().jwtSecret)}`

const createValidTokenHeader = () =>
  `Bearer ${sign(<MockToken>{ userId: 'foo', role: Roles.User }, getConfig().jwtSecret)}`

const createEnforceProviderThatReturnsAuthorized = () => TE.right(() => TE.right(true))
const createEnforceProviderThatReturnsUnauthorized = () => TE.right(() => TE.right(false))
const createEnforceProviderThatFailsToEnforce = () =>
  TE.right(() => TE.left({ code: 'ENFORCER_FAILED' }))
const createEnforceProviderThatFails = () => TE.left({ code: 'CANNOT_GET_ENFORCER' })

describe('enforceWithAuthHeader middleware', () => {
  describe.each<[string, string, EnforceProvider]>([
    [
      'when enforcer returns authorized and token is valid',
      createValidTokenHeader(),
      createEnforceProviderThatReturnsAuthorized(),
    ],
  ])('%s', (_, token, enforceProvider) => {
    const { mockNext, mockContext } = createMockContextAndNext(token)
    let logSpy: jest.SpyInstance

    beforeAll(async () => {
      const logger = createMockLogger()
      logSpy = jest.spyOn(logger, 'log')
      await createEnforceWithAuthHeaderMiddleware(
        enforceProvider,
        getConfig,
        () => logger,
        mockTokenV,
        getRoleFromToken,
      )('task', ['add'])(mockContext, mockNext)
    })

    test('should not set status to 401', () => {
      expect(mockContext.status).not.toEqual(HttpStatus.UNAUTHORIZED)
    })

    test('should invoke next once', () => {
      expect(mockNext).toHaveBeenCalledTimes(1)
    })

    test('should log error', () => {
      expect(logSpy.mock.calls).toMatchSnapshot()
    })
  })

  describe.each<[string, string, EnforceProvider]>([
    [
      'when enforcer returns unauthorized and token is valid',
      createValidTokenHeader(),
      createEnforceProviderThatReturnsUnauthorized(),
    ],
    [
      'when enforcer returns unauthorized and token schema is invalid',
      createInvalidTokenSchemaHeader(),
      createEnforceProviderThatReturnsUnauthorized(),
    ],
    ['when enforcer fails', createValidTokenHeader(), createEnforceProviderThatFailsToEnforce()],
    ['when cannot get enforcer', createValidTokenHeader(), createEnforceProviderThatFails()],
  ])('%s', (_, token, enforceProvider) => {
    const { mockNext, mockContext } = createMockContextAndNext(token)
    let logSpy: jest.SpyInstance

    beforeAll(async () => {
      const logger = createMockLogger()
      logSpy = jest.spyOn(logger, 'log')
      await createEnforceWithAuthHeaderMiddleware(
        enforceProvider,
        getConfig,
        () => logger,
        mockTokenV,
        getRoleFromToken,
      )('task', ['add'])(mockContext, mockNext)
    })

    test('should set status to 401', () => {
      expect(mockContext.status).toEqual(HttpStatus.UNAUTHORIZED)
    })

    test('should not invoke next', () => {
      expect(mockNext).not.toHaveBeenCalled()
    })

    test('should log error', () => {
      expect(logSpy.mock.calls).toMatchSnapshot()
    })
  })

  describe.each<[string, string]>([
    ['not using Bearer protocol', `Basic foo`],
    ['header has an invalid token format', 'Bearer foo'],
    ['token schema is invalid', createInvalidTokenSchemaHeader()],
    ['header has too many parts', `${createValidTokenHeader()} bar`],
    ['header is missing', undefined as any],
    ['token is expired', createExpiredTokenHeader()],
  ])('when enforcer returns authorized and %s', (_, authHeader) => {
    const mockContext = createKoaContext({
      headers: { authorization: authHeader },
    })
    const mockNext = createMockNext()
    let logSpy: jest.SpyInstance

    beforeAll(async () => {
      const logger = createMockLogger()
      logSpy = jest.spyOn(logger, 'log')
      await createEnforceWithAuthHeaderMiddleware(
        createEnforceProviderThatReturnsAuthorized(),
        getConfig,
        () => logger,
        mockTokenV,
        getRoleFromToken,
      )('task', ['add'])(mockContext, mockNext)
    })

    test('should set status to 401', () => {
      expect(mockContext.status).toEqual(HttpStatus.UNAUTHORIZED)
    })

    test('should not invoke next', () => {
      expect(mockNext).not.toHaveBeenCalled()
    })

    test('should log error', () => {
      expect(logSpy.mock.calls).toMatchSnapshot()
    })
  })
})
