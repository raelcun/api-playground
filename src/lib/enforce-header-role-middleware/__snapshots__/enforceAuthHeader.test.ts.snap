// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enforceWithAuthHeader middleware when cannot get enforcer should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "CANNOT_GET_ENFORCER",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer fails should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "ENFORCER_FAILED",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns authorized and header has an invalid token format should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "INVALID_TOKEN",
        "message": "jwt malformed",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns authorized and header has too many parts should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "INVALID_AUTH_HEADER",
        "message": "requires two parts",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns authorized and header is missing should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "INVALID_AUTH_HEADER",
        "message": "invalid auth header format",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns authorized and not using Bearer protocol should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "INVALID_AUTH_HEADER",
        "message": "first part must be \\"Bearer\\"",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns authorized and token is expired should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "INVALID_TOKEN",
        "message": "jwt expired",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns authorized and token is valid should log error 1`] = `Array []`;

exports[`enforceWithAuthHeader middleware when enforcer returns authorized and token schema is invalid should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "INVALID_TOKEN",
        "message": "token payload format invalid",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns unauthorized and token is valid should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "ROLE_REJECTED",
      },
    },
  ],
]
`;

exports[`enforceWithAuthHeader middleware when enforcer returns unauthorized and token schema is invalid should log error 1`] = `
Array [
  Array [
    Object {
      "level": "error",
      "payload": Object {
        "code": "INVALID_TOKEN",
        "message": "token payload format invalid",
      },
    },
  ],
]
`;
